### ISSUE TITLE
Nullable struct converter

### ISSUE THREAD

**Issue by fleed** [dotnet/corefx#40229]
I've written the following converter to consume data from a 3rdparty API:

```csharp
public class JsonNullableDateTimeOffsetConverter : JsonConverter<DateTimeOffset?>
    {
        public override DateTimeOffset? Read(ref Utf8JsonReader reader, Type typeToConvert,
            JsonSerializerOptions options)
        {
            var value = reader.GetString();
            if (string.IsNullOrEmpty(value))
            {
                return default;
            }

            return DateTimeOffset.ParseExact(reader.GetString(), "yyyy/MM/dd HH:mm:ss zzz",
                CultureInfo.InvariantCulture);
        }

        public override void Write(Utf8JsonWriter writer, DateTimeOffset? value, JsonSerializerOptions options)
        {
            if (!value.HasValue)
            {
                return;
            }
            
            writer.WriteStringValue(value.Value.ToString("yyyy/MM/dd HH:mm:ss zzz"));
        }
    }
```

The following test

```csharp
[Fact]
        public void TestNullableDateTimeOffset()
        {
            var s = "{\"nullableValue\":null}";
            var v = JsonSerializer.Deserialize<TestNullable>(s);
        }
```

throws the exception:
**System.FormatException : Index (zero based) must be greater than or equal to zero and less than the size of the argument list.**

I have the same exception also when selecting a converter that doesn't match the type.
I think that at least a more meaningful exception should be thrown.

**Comment by steveharter** [comment]
@fleed I ran this against the latest code which did not throw. The class `TestNullable` wasn't provided above so I used:
```
    public class TestNullable
    {
        public DateTimeOffset? nullableValue { get; set; }
    }
```

Can you verify the test class above and also try the latest 3.0.x [runtime
](https://github.com/dotnet/core-setup#daily-builds)

**Comment by steveharter** [comment]
The FormatException is an [issue](https://github.com/dotnet/corefx/pull/39789) that was recently fixed. You can install the latest 3.0.x [runtime ](https://github.com/dotnet/core-setup#daily-builds)to see that fix.

The actual message should be InvalidOperationException "The converter specified on 'xyz' is not compatible with the type 'System.DateTimeOffset'."

The `TestNullable` class was not provided with your repro. I assume that it should be:
```cs
public class TestNullable
{
    [JsonConverter(typeof(JsonNullableDateTimeOffsetConverter))]
    public DateTimeOffset? nullableValue { get; set; }
 }
```

The deserializer automatically handles null values, meaning object and `Nullable<T>` custom converters do not need to handle null with the normal semantics (where "null" in json means a null object or nullable).

However it appears you want to change that to set the property to return null for an empty DateTimeOffset given an empty JSON string (""). However, that is currently not supported due to the default null handling explained earlier (it would work for objects, but not `Nullable<T>`).

I will investigate further and report back.

**Comment by fleed** [comment]
@steveharter I tested it using the preview8 version and I get the `InvalidOperationException` that you mentioned.

**Comment by ahsonkhan** [comment]
Re-opening until fixed in 3.0

**Comment by steveharter** [comment]
Fixed in 3.0 via https://github.com/dotnet/corefx/pull/40357

**Comment by fleed** [comment]
@steveharter Thank you. I can confirm that it's fixed in **preview 9**
